### ISSUE TITLE
Publish target makes writes to intermediate folder, potentially breaking parallel builds. 

### ISSUE THREAD

**Issue by aaronla-ms** [dotnet/sdk#16586]
The Publish target gathers files first to a location specificed by PublishTmp. This is a problem, as sfproj and other service packagers might invoke Publish in parallel with different public locations. 

Repro: see repro attached to https://github.com/dotnet/msbuild/issues/6309

If you invoke Package target on both Application1 and Application2 in parallel, a race condition can occur on the PublishTmp folder.

**Comment by sfoslund** [comment]
@aaronla-ms are you able to write to separate tmp directories in this case? You can try specifying different tmp dirs in your project files.

**Comment by aaronla-ms** [comment]
@sfoslund currently we work around this by hashing the $(PublishUrl) variable to set $(PublishIntermediateOutputPath) to a unique location.

```
<Project>
  <!--
    For projects published with DotNetPublish, by default that target will publish into a temp folder first. Unfortunately, that
    can result in race conditions if two sfproj's attempt to DotNetPublish the same csproj at the same time, as both runs of the csproj
    DotNetPublish will use the same temp folder.

    It is observed that DotNetPublish is always invoked with a unique PublishUrl.

    To mitigate this issue, when PublishUrl set, the PublishIntermediateOutputPath will be set derived from a hash of PublishUrl.
  -->
  <Target Name="FixupPublishIntermediateOutputPath" Condition="'$(PublishUrl)'!=''" BeforeTargets="$(_DotNetPublishCopyFiles);_PrepareForAspNetCoreProjectSystemPublish">
    <!-- OutLength=20 (120 bit) should provide the same collision resistance as a guid, while avoiding hitting maxpath limits. -->
    <HashString InString="$(PublishUrl)" OutLength="20">
      <Output TaskParameter="OutString" PropertyName="PublishUrl_Hash"/>
    </HashString>
    <Message Text="Choosing unique PublishIntermediateOutputPath for publish-during-sfproj." />
    <Message Text="PublishIntermediateOutputPath was '$(PublishIntermediateOutputPath)'" />
    <PropertyGroup>
      <PublishIntermediateTempPath>$(PublishIntermediateTempPath.Replace('\PubTmp\', '\PubTmp-$(PublishUrl_Hash)\'))</PublishIntermediateTempPath>
      <PublishIntermediateOutputPath>$(PublishIntermediateOutputPath.Replace('\PubTmp\', '\PubTmp-$(PublishUrl_Hash)\'))</PublishIntermediateOutputPath>
    </PropertyGroup>
    <Message Text="PublishIntermediateOutputPath is now '$(PublishIntermediateOutputPath)'" />
  </Target>

  <!-- Custom Tasks -->
  <UsingTask
    TaskName="HashString"
    TaskFactory="$(CodeTaskFactory_TaskFactory)"
    AssemblyName="$(CodeTaskFactory_AssemblyName)">
    <ParameterGroup>
      <InString ParameterType="System.String" Required="true" />
      <OutLength ParameterType="System.Int32" />
      <OutString ParameterType="System.String" Output="true" />
    </ParameterGroup>
    <Task>
      <Using Namespace="System"/>
      <Using Namespace="System.Text"/>
      <Using Namespace="System.Security.Cryptography"/>
      <Code Type="Fragment" Language="cs">
        <![CDATA[
        using (SHA256 sha256Hash = SHA256.Create())
        {
          var result = Convert.ToBase64String(sha256Hash.ComputeHash(Encoding.UTF8.GetBytes(InString)));
          // truncate to caller preference
          if (OutLength > 0 && result.Length > OutLength)
          {
            result = result.Substring(0, OutLength);
          }
          // make file path safe
          result = result.Replace("+", "-").Replace("/", "_").Replace("=", "");
          OutString = result;
        }
        ]]>
      </Code>
    </Task>
  </UsingTask>
</Project>
```

**Comment by sfoslund** [comment]
That seems like a good approach. @dsplaisted is there anything simpler that can be done here?

**Comment by aaronla-ms** [comment]
Thanks @sfoslund. Though, ideally we'd want to compute the temp directory during project import time, before any targets run. Unfortunately, I couldn't find any clean way to extend MSBuild expressions with a HashString function that could be invoked during import time. Perhaps you know of a way to do so.

**Comment by dsplaisted** [comment]
I tried to look at the repro but I get an error that there's not `Package` target:

> error MSB4057: The target "Package" does not exist in the project.

**Comment by aaronla-ms** [comment]
@dsplaisted did you remember to restore nuget packages for all the sfproj projects? Remember that they don't use `dotnet restore` to restore

**Comment by aaronla-ms** [comment]
For those not familiar with packages.config:

1. dotnet restore Stateless1
2. nuget restore -ConfigFile nuget.config -PackagesDirectory packages Application1\packages.config
3. dotnet msbuild Application1 /t:Package

**Comment by dsplaisted** [comment]
OK, I restored using NuGet.exe, but I'm not seeing any reference to a `PublishTmp` folder in the binlog when I run `msbuild /t:Package Application1 /p:BuildProjectReferences=false /bl:Application1.binlog`

**Comment by aaronla-ms** [comment]
Ah, you're right. My Mistake. One more change:

1. Edit Stateless1.csproj, change `Sdk="Microsoft.NET.Sdk"` to `Sdk="Microsoft.NET.Sdk.Web"`
2. search for PubTmp or PublishIntermediateTempPath in the binlog.

You should see something like this:

> Target _CopyResolvedFilesToPublishPreserveNewest
> Building target "_CopyResolvedFilesToPublishPreserveNewest" completely.
> ...
> Task Copy
> ...
> Copy file from "...obj\Debug\net462\win7-x64\Stateless1.exe" to "...obj\Debug\net462\win7-x64\PubTmp\Out\Stateless1.exe"

And so on.

**Comment by dsplaisted** [comment]
Generally it's not safe to build the same project in parallel with different global properties that don't affect the output paths.

I'm not sure if we should try to change the default `PublishIntermediateTempPath` for this.  It would be nice to have a better way to generate a hash of a value during MSBuild evaluation.  @BenVillalobos @cdmihai @jeffkl

**Comment by aaronla-ms** [comment]
@dsplaisted ok, fair enough, but then how should sfproj projects consume such services?

E.g. if publishing happened during build, then sfproj wouldn't need to invoke publish potentially in parallel with other projects. This is how sfproj consumes old-style c# project for instance -- it just enumerates the outputs of the project.\

Unfortunately, for Microsoft.NET.Sdk.Web project, the outdir folder doesn't contain the necessary parts.

**Comment by aaronla-ms** [comment]
Ah, I see that now -- PublishUrl defaults to $(OutDir)\Publish. I'll sync up with the service fabric team to see if they could do that instead in sfprojs.

In that case, would we have to modify the service to publish during build? E.g.

```
<Project Sdk="Microsoft.NET.Sdk.Web">
  <PropertyGroup>
    <!-- Tell publish not to invoke Build target. Otherwise, circular dependency between targets error -->
    <NoBuild>true</NoBuild>
  </PropertyGroup>
  <Target Name="PublishAfterBuild" BeforeTargets="AfterBuild" DependsOnTargets="Publish" />
</Project>
```

Is that what you're suggesting?

**Comment by dsplaisted** [comment]
> In that case, would we have to modify the service to publish during build?

How do you consume old-style c# projects?  Could you do something similar but call Publish instead of Build?

**Comment by aaronla-ms** [comment]
> How do you consume old-style c# projects? Could you do something similar but call Publish instead of Build?

Building referenced projects already takes care of that. ResolveProjectReferences invokes the Build target on referenced project, not the Publish target.

It might help to set up an sfproj with both a Publish style project and an old style project. For the old style project, sfproj is already coded to just copy the outdir -- it doesn't invoke any targets recursively, as ResolveProjectReferences already makes sure that all referenced Projects are already built.
